app/models/piece.rb

```diff
@@ -79,21 +65,83 @@ def moved?(x, y)
   end
 
   # Returns true if the coordinates provided
-  # have the same x-axis value.
-  def horizontal_move?(x)
-    x_distance(x) == 0
+  # have the same x-axis value and there are no
+  # pieces in between.
+  def clear_horizontal_move?(x, y)
+    return false unless y_distance(y) == 0
+    distance = x_distance(x)
+    path_clear?(x, y, distance)
   end
 
   # Returns true if the coordinates provided
-  # have the same y-axis value.
-  def vertical_move?(y)
-    y_distance(y) == 0
+  # have the same y-axis value and there are no
+  # pieces in between.
+  def clear_vertical_move?(x, y)
+    return false unless x_distance(x) == 0
+    distance = y_distance(y)
+    path_clear?(x, y, distance)
   end
 
   # Returns true if the coordinates provided
   # are the same distance away from the
-  # origin point along both axis.
-  def diagonal_move?(x, y)
-    x_distance(x) == y_distance(y)
+  # origin point along both axis and there are no
+  # pieces in between.
+  def clear_diagonal_move?(x, y)
+    return false unless x_distance(x) == y_distance(y)
+    distance = x_distance(x)
+    path_clear?(x, y, distance)
+  end
+
+  # Class constants used to determine path direction
+  # for obstruction logic.
+  RIGHT = 1
```

Comment:
Above. I may be slightly off in regards to Rails styling, but here's a paste out of LasLabs styling guide:

In a Model attribute order should be

1. Private attributes (_name, _description, _inherit, ...)
2. Class constants
3. Fields declarations
4. Default method and _default_get
5. Compute, inverse, search, and group methods in the same order as field declaration
6. Constrains methods (@api.constrains) and onchange methods (@api.onchange)
7. CRUD methods (ORM overrides)
8. Action methods
9. And finally, other business methods.

You don't have 1, 4, and 6 for the most part.